chain/client/src/sync/state/shard.rs

```diff
@@ -0,0 +1,268 @@
+    assert!(flat_storage_manager.get_flat_storage_for_shard(shard_uid).is_none());
+
+    let flat_head_hash = *chunk.prev_block();
+    let flat_head_header =
+        store.get_ser::<BlockHeader>(DBCol::BlockHeader, flat_head_hash.as_bytes())?.ok_or_else(
+            || near_chain::Error::DBNotFoundErr(format!("No block header {}", flat_head_hash)),
+        )?;
+    let flat_head_prev_hash = *flat_head_header.prev_hash();
+    let flat_head_height = flat_head_header.height();
+
+    tracing::debug!(target: "store", ?shard_uid, ?flat_head_hash, flat_head_height, "set_state_finalize - initialized flat storage");
+
+    let mut store_update = store.flat_store().store_update();
+    store_update.set_flat_storage_status(
+        shard_uid,
+        FlatStorageStatus::Ready(FlatStorageReadyStatus {
+            flat_head: near_store::flat::BlockInfo {
+                hash: flat_head_hash,
+                prev_hash: flat_head_prev_hash,
+                height: flat_head_height,
+            },
+        }),
+    );
+    store_update.commit()?;
+    flat_storage_manager.create_flat_storage_for_shard(shard_uid).unwrap();
+    Ok(())
+}
+
+async fn apply_state_part(
+    store: Store,
+    runtime: Arc<dyn RuntimeAdapter>,
+    computation_task_tracker: TaskTracker,
+    cancel: CancellationToken,
+    sync_hash: CryptoHash,
+    shard_id: ShardId,
+    part_id: u64,
+    num_parts: u64,
+    state_root: CryptoHash,
+    epoch_id: EpochId,
+) -> anyhow::Result<(), near_chain::Error> {
+    return_if_cancelled!(cancel);
+    let handle =
+        computation_task_tracker.get_handle(&format!("shard {} part {}", shard_id, part_id)).await;
+    return_if_cancelled!(cancel);
+    handle.set_status("Loading part data from store");
+    let data = store
+        .get(
+            DBCol::StateParts,
+            &borsh::to_vec(&StatePartKey(sync_hash, shard_id, part_id)).unwrap(),
+        )?
+        .ok_or_else(|| {
+            near_chain::Error::DBNotFoundErr(format!(
+                "No state part {} for shard {}",
+                part_id, shard_id
+            ))
+        })?
+        .to_vec();
+    handle.set_status("Applying part data to runtime");
+    runtime.apply_state_part(
+        shard_id,
+        &state_root,
+        PartId { idx: part_id, total: num_parts },
+        &data,
+        &epoch_id,
+    )?;
+    Ok(())
+}
+
+/// Given a future, respawn it as an equivalent future but which does not block the
+/// driver of the future. For example, if the given future directly performs
+/// computation, normally the whoever drives the future (such as a buffered_unordered)
+/// would be blocked by the computation, thereby not allowing computation of other
+/// futures driven by the same driver to proceed. This function respawns the future
+/// onto the FutureSpawner, so the driver of the returned future would not be blocked.
+fn respawn_for_parallelism<T: Send + 'static>(
```

Comment:
Instead of this, is it not possible to have small enough compute intensive tasks so as to not block the driver? This looks a bit odd overall where we are effectively "transferring" the future to a different spawner.awkward

Reply:
That's basically what it does; the thing is that tokio_stream::iter never spawns anything; it only awaits multiple futures on the same driver.